src/metadata/metadata_io.c

```diff
@@ -177,147 +177,184 @@ int metadata_io_read_i_atomic(ocf_cache_t cache, ocf_queue_t queue, void *priv,
-	ocf_cache_t cache;
-
-	OCF_CHECK_NULL(request);
+	struct metadata_io_request_asynch *a_req = m_req->asynch;
+	ocf_cache_t cache = m_req->cache;
 
-	cache = request->cache;
-
-	a_req = request->asynch;
 	OCF_CHECK_NULL(a_req);
 	OCF_CHECK_NULL(a_req->on_complete);
 
 	if (error) {
-		request->error |= error;
-		request->asynch->error |= error;
+		a_req->error = a_req->error ?: error;
 	} else {
-		if (request->fl_req.rw == OCF_READ)
-			metadata_io_req_drain(request);
+		if (m_req->req.rw == OCF_READ)
+			metadata_io_req_drain(m_req);
 	}
 
-	if (env_atomic_dec_return(&request->req_remaining))
-		return;
+	OCF_DEBUG_PARAM(cache, "Page = %u", m_req->page);
 
-	OCF_DEBUG_PARAM(cache, "Page = %u", request->page);
+	metadata_io_req_advance(m_req);
 
-	ctx_data_free(cache->owner, request->data);
-	request->data = NULL;
+	env_atomic_set(&m_req->finished, 1);
+	ocf_metadata_updater_kick(cache);
+}
 
-	if (env_atomic_dec_return(&a_req->req_remaining)) {
-		env_atomic_set(&request->finished, 1);
-		ocf_metadata_updater_kick(cache);
-		return;
-	}
+static void metadata_io_req_submit(struct metadata_io_request *m_req)
+{
+	env_atomic_set(&(m_req->finished), 0);
+	metadata_updater_submit(m_req);
+}
 
-	OCF_DEBUG_MSG(cache, "Asynchronous IO completed");
+void metadata_io_req_end(struct metadata_io_request *m_req)
+{
+	struct metadata_io_request_asynch *a_req = m_req->asynch;
+	ocf_cache_t cache = m_req->cache;
 
-	/* All IOs have been finished, call IO end callback */
-	a_req->on_complete(request->cache, a_req->context, request->error);
+	if (env_atomic_dec_return(&a_req->req_remaining) == 0)
+		a_req->on_complete(cache, a_req->context, a_req->error);
 
-	/*
-	 * If it's last request, we mark is as finished
-	 * after calling IO end callback
-	 */
-	env_atomic_set(&request->finished, 1);
-	ocf_metadata_updater_kick(cache);
+	ctx_data_free(cache->owner, m_req->data);
+}
+
+void metadata_io_req_finalize(struct metadata_io_request *m_req)
+{
+	struct metadata_io_request_asynch *a_req = m_req->asynch;
+
+	if (env_atomic_dec_return(&a_req->req_active) == 0)
+		env_free(a_req);
 }
 
 static uint32_t metadata_io_max_page(ocf_cache_t cache)
 {
 	return ocf_volume_get_max_io_size(&cache->device->volume) / PAGE_SIZE;
 }
 
-static void metadata_io_req_error(ocf_cache_t cache,
-				  struct metadata_io_request_asynch *a_req,
-				  uint32_t i, int error)
+static void metadata_io_req_advance(struct metadata_io_request *m_req)
 {
-	a_req->error |= error;
-	a_req->reqs[i].error |= error;
-	a_req->reqs[i].count = 0;
-	if (a_req->reqs[i].data)
-		ctx_data_free(cache->owner, a_req->reqs[i].data);
-	a_req->reqs[i].data = NULL;
+	struct metadata_io_request_asynch *a_req = m_req->asynch;
+	uint32_t max_count = metadata_io_max_page(m_req->cache);
+	uint32_t curr;
+
+	if (a_req->error) {
+		metadata_io_req_end(m_req);
+		return;
+	}
+
+	curr = env_atomic_inc_return(&a_req->req_current);
+
+	if (curr >= OCF_DIV_ROUND_UP(a_req->count, max_count)) {
```

Comment:
I think this should be '>':
if (curr > OCF_DIV_ROUND_UP(a_req->count, max_count)) {

Reply:
The `curr` is counted from 0, so we maximum `curr` should be io count - 1.